Clover/app/src/main/java/org/floens/chan/core/manager/PageRequestManager.java

```diff
@@ -0,0 +1,153 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.floens.chan.core.manager;
+
+import org.floens.chan.core.model.Post;
+import org.floens.chan.core.model.orm.Board;
+import org.floens.chan.core.model.orm.Loadable;
+import org.floens.chan.core.site.Page;
+import org.floens.chan.core.site.Pages;
+import org.floens.chan.core.site.SiteActions;
+import org.floens.chan.core.site.ThreadTime;
+import org.floens.chan.utils.Logger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class PageRequestManager implements SiteActions.PagesListener {
+    private static final String TAG = "PageRequestManager";
+
+    private Set<String> requestedBoards = Collections.synchronizedSet(new HashSet<>());
+    private Set<String> savedBoards = Collections.synchronizedSet(new HashSet<>());
+    private ConcurrentMap<String, Pages> boardPagesMap = new ConcurrentHashMap<>();
+    private ConcurrentMap<String, Long> boardTimeMap = new ConcurrentHashMap<>();
+
+    private List<PageCallback> callbackList = new ArrayList<>();
+
+    @Inject
+    public PageRequestManager() {
+    }
+
+    public Page getPage(Post op) {
+        if (op == null) {
+            return null;
+        }
+        Pages pages = getPages(op.board);
+        if (pages == null) {
+            return null;
+        } else {
+            for (Page page : pages.pages) {
+                for (ThreadTime threadTime : page.threads) {
+                    if (op.no == threadTime.no) {
+                        return page;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    public Page getPage(Loadable opLoadable) {
+        if (opLoadable == null) {
+            return null;
+        }
+        Pages pages = getPages(opLoadable.board);
+        if (pages == null) {
+            return null;
+        } else {
+            for (Page page : pages.pages) {
+                for (ThreadTime threadTime : page.threads) {
+                    if (opLoadable.no == threadTime.no) {
+                        return page;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    private Pages getPages(Board b) {
+        if (savedBoards.contains(b.code)) {
+            //if we have it stored already, return the pages for it
+            //also issue a new request if 3 minutes have passed
+            long lastUpdateTime = boardTimeMap.get(b.code);
+            if (lastUpdateTime + 3 * 1000 * 60 > System.currentTimeMillis()) {
```

Comment:
you could flip this boolean and the else into the if block, to avoid empty blocks

Reply:
maybe constantize the `3 * 1000 * 60` (`THREE_MINUTES`?) and/or wrap this check in a method that could be called  something like `shouldThreadUpdate(lastUpdateTime)`

Reply:
I can implement both of those.